File: Labs/week4/A_string_matching.py
def solve():
    n = int(input())
    pro = {'<': 0, '(': 1, '[': 2, '{': 3}
    zu = {'>': '<', ')': '(', ']': '[', '}': '{'}
    for _ in range(n):
        f = input().strip()
        s = []
        ok = True
        for ch in f:
            if ch in pro:
                if s and pro[ch] > pro[s[-1]]:
                    ok = False
                    break
                s.append(ch)
            elif ch in zu:
                if not s or zu[ch] != s[-1]:
                    ok = False
                    break
                s.pop()
        print("YES" if ok and not s else "NO")

if __name__ == "__main__":
    solve()

File: Labs/week4/A_string_matching.cpp
#include <iostream>
#include <stack>
#include <map>
using namespace std;

int main() {
    int n;
    cin >> n;
    getchar();
    map<char, int> pro = {{'<', 0}, {'(', 1}, {'[', 2}, {'{', 3}};
    map<char, char> zu = {{'>', '<'}, {')', '('}, {']', '['}, {'}', '{'}};

    for (int i = 0; i < n; i++) {
        stack<char> s;
        string f;
        getline(cin, f);
        bool ok = true;
        for (char ch : f) {
            if (pro.count(ch)) {
                if (!s.empty() && pro[ch] > pro[s.top()]) {
                    ok = false;
                    break;
                }
                s.push(ch);
            } else if (zu.count(ch)) {
                if (s.empty() || zu[ch] != s.top()) {
                    ok = false;
                    break;
                }
                s.pop();
            }
        }
        cout << (ok && s.empty() ? "YES" : "NO") << endl;
    }
    return 0;
}
